server/namespace_cluster.go

```diff
@@ -0,0 +1,128 @@
+// limitations under the License.
+
+package server
+
+import (
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/namespace"
+	"github.com/pingcap/pd/server/schedule"
+)
+
+// namespaceCluster is part of a global cluster that contains stores and reigons
+// within a specific namespace.
+type namespaceCluster struct {
+	schedule.Cluster
+	classifier namespace.Classifier
+	namespace  string
+	stores     map[uint64]*core.StoreInfo
+}
+
+func newNamespaceCluster(c schedule.Cluster, classifier namespace.Classifier, namespace string) *namespaceCluster {
+	stores := make(map[uint64]*core.StoreInfo)
+	for _, s := range c.GetStores() {
+		if classifier.GetStoreNamespace(s) == namespace {
+			stores[s.GetId()] = s
+		}
+	}
+	return &namespaceCluster{
+		Cluster:    c,
+		classifier: classifier,
+		namespace:  namespace,
+		stores:     stores,
+	}
+}
+
+func (c *namespaceCluster) checkRegion(region *core.RegionInfo) bool {
+	for _, p := range region.Peers {
+		if _, ok := c.stores[p.GetStoreId()]; !ok {
+			return false
+		}
+	}
+	return true
+}
+
+const randRegionMaxRetry = 10
+
+// RandFollowerRegion returns a random region that has a follower on the store.
+func (c *namespaceCluster) RandFollowerRegion(storeID uint64) *core.RegionInfo {
+	for i := 0; i < randRegionMaxRetry; i++ {
+		r := c.Cluster.RandFollowerRegion(storeID)
+		if r == nil {
+			return nil
+		}
+		if c.checkRegion(r) {
+			return r
+		}
+	}
+	return nil
+}
+
+// RandLeaderRegion returns a random region that has leader on the store.
+func (c *namespaceCluster) RandLeaderRegion(storeID uint64) *core.RegionInfo {
+	for i := 0; i < randRegionMaxRetry; i++ {
+		r := c.Cluster.RandLeaderRegion(storeID)
+		if r == nil {
+			return nil
+		}
+		if c.checkRegion(r) {
+			return r
+		}
+	}
+	return nil
+}
+
+// GetStores returns all stores in the namespace.
+func (c *namespaceCluster) GetStores() []*core.StoreInfo {
+	stores := make([]*core.StoreInfo, 0, len(c.stores))
+	for _, s := range c.stores {
+		stores = append(stores, s)
+	}
+	return stores
+}
+
+// GetStore searches for a store by ID.
+func (c *namespaceCluster) GetStore(id uint64) *core.StoreInfo {
+	return c.stores[id]
+}
+
+// GetRegion searches for a region by ID.
+func (c *namespaceCluster) GetRegion(id uint64) *core.RegionInfo {
+	r := c.Cluster.GetRegion(id)
+	if r == nil || !c.checkRegion(r) {
+		return nil
+	}
+	return r
+}
+
+// RegionWriteStats returns hot region's write stats.
+func (c *namespaceCluster) RegionWriteStats() []*core.RegionStat {
+	allStats := c.Cluster.RegionWriteStats()
+	stats := make([]*core.RegionStat, 0, len(allStats))
+	for _, s := range allStats {
+		if c.GetRegion(s.RegionID) != nil {
+			stats = append(stats, s)
+		}
+	}
+	return stats
+}
+
+func scheduleByNamespace(cluster schedule.Cluster, classifier namespace.Classifier, scheduler schedule.Scheduler) *schedule.Operator {
+	for _, ns := range classifier.GetAllNamespaces() {
```

Comment:
Is there any chance that the region on the beginning of the the namespace list got scheduled, while the ones on the end of the list not

Reply:
Yes, it's possible. I'll change it to pick namespace in random order.

Reply:
LGTM